Athena cloud formation (#5)
quiltdata/examples: athena_cfn/README.md

10. Click Next
11. Click Next
12. At the bottom, open the "Quick-create link" triangle
13. Click "**Open quick-create link**" to get a wizard that makes it easy to
rerun this for multiple buckets.

Save this link in a shared space so you and other AWS users can easily index additional Quilt buckets.

### B. Use quick-create link for each bucket you want to index

1. Open the link from (A.13)
2. Update the stack name and bucket parameters to match THIS bucket
3. Click "Create change set"
4. Wait until "Status = CREATE_COMPLETE" (may need to refresh using upper-right "cycle" icon)
5. Review Changes
6. Click "Execute" in upper right
7. Select "Roll back all stack resources" (default)
8. Click "Execute change set"

### C. Manually invoke view-creating queries

Unfortunately, CloudFormation does not automatically generate the proper views.
To complete setup, you will need to manually execute two pre-defined queries for each bucket:

1. Login to your Quilt Repository
2. Go to the "Queries" tab
3. Click "Select workgroup" (matches the stack name you chose)
4. Click "Select query"
5. Find the "create..." queries corresponding to this bucket

> 5. Select "create1...quilt_packages_view" and click "Run Query"

Author: Fair enough.  Makes it easier to be consistent.

# Query Quilt Package Metadata in AWS Athena

The code in this example shows you how to enable the "Queries"
tab of the Quilt Catalog for querying Quilt [metadata](https://docs.quiltdata.com/catalog/metadata)
in SQL with [Quilt's Athena Integration](https://docs.quiltdata.com/advanced/athena).

## I. Prologue: Ensure Source=Quilt Roles for Athena users

Quilt stacks ship with two `Source=Custom` roles that
automatically grant access to all buckets that have been added to Quilt.
At this time `Source=Custom` Roles are incompatible with custom policies
(including Athena policies). If you have not already,
before proceeding you must do the following in the Quilt Admin panel:

> 1. Create a new Quilt Role (`Source=Quilt`), such as "AthenaQuiltAccess"

Author: It does not. See below.

> 1. Create a new Quilt Policy (`Source=Quilt`) with access to any

Reviewer: Ah, now I know why this is confusing. We need to explain way up front that `Source=Quilt` means check the box "Manually set ARN instead of configuring policies" and provide your own policy ARN.

Author: Um, I think that is the other way around.  `Source=Custom` means manually-set ARNs, `Source=Quilt` means configured directly from the UI.

I agree that https://docs.quiltdata.com/catalog/admin needs to be MUCH clearer on this, but I wanted to decouple that from this PR.

# Query Quilt Package Metadata in AWS Athena

> The code in this example shows you how to enable use of the "Queries"

Reviewer: ```suggestion
The code in this example shows you how to enable the "Queries"
```

> tab of the Quilt Catalog for querying Quilt [metadata](https://docs.quiltdata.com/catalog/metadata) for a given bucket using [Athena SQL](https://docs.quiltdata.com/advanced/athena).

Reviewer: ```suggestion
tab of the Quilt Catalog for querying Quilt [metadata](https://docs.quiltdata.com/catalog/metadata)
in SQL with [Quilt's Athena Integration](https://docs.quiltdata.com/advanced/athena).
```

> ### Validating CloudFormation files

Reviewer: This section, for readability, should probably be bumped much later, when the customer might actually generate the templates. Also we should remind customers to switch to their desired environment before `pip install` etc.

Author: Moved

If for any reason you need to edit the CloudFormation files,

> be sure to lint them before using:

Reviewer: ```suggestion
we recommend that you lint them as follows:
```

```
$ pip3 install --upgrade taskcat cfn-lint
$ taskcat lint && cfn-lint *cfn.yml
```

## I. Prologue: Ensure Source=Quilt Roles for Athena users

> The initial Quilt deployment ships with two `Source=Custom` Roles that

Reviewer: ```suggestion
Quilt stacks ship with two `Source=Custom` roles that
```

1. Select "Amazon S3 URL" (default)
1. Paste in the `athena-bucket-cfn.yml` Object URL from before
1. Click Next
1. Enter a new "Stack name", e.g.: `your-bucket-athena-quilt`
1. Enter "QuiltAthenaStack", e.g.: `athena-quilt` (from part III.A.7)
1. Enter "QuiltBucket" to pull metadata from, e.g.: `your-bucket` (without `s3` or `arn` prefixes)
1. Enter "QuiltBucketID", e.g.: `your_bucket` (with '-' replaced by '_')
1. Click Next
1. Click Next
1. At the bottom, open the "Quick-create link" triangle
1. Click "**Open quick-create link**" to get a wizard that makes it easy to
rerun this for multiple buckets.

Save this link in a shared space so you and other AWS users can easily index additional Quilt buckets.

### B. Use quick-create link for each bucket you want to index

1. Open the link from (A.13)
1. Update the stack name and bucket parameters to match THIS bucket
1. Click "Create change set"
1. Wait until "Status = CREATE_COMPLETE" (may need to refresh using upper-right "cycle" icon)
1. Review Changes
1. Click "Execute" in upper right
1. Select "Roll back all stack resources" (default)
1. Click "Execute change set"

### C. Manually invoke view-creating queries

> Unfortunately, CloudFormation does not automatically generate the proper views.

Reviewer: I think we should say that this is "overly complicated" rather than impossible? https://stackoverflow.com/questions/56289272/create-aws-athena-view-programmatically#56347331

Reviewer: It's also outside the scope of the spirit of CloudFormation which is infra, not so much business logic like views.

# Query Quilt Package Metadata in AWS Athena

> The code in this example shows you to enable (and test) use of the "Queries"

Reviewer: ```suggestion
The code in this example shows you to enable the "Queries"
```

> pane of the Web Catalog for querying Quilt metadata for specific S3 bucket.

Reviewer: ```suggestion
tab of the Quilt Catalog for querying package metadata for a given bucket.
```

> ### NOTE: Validating CloudFormation files

Reviewer: ```suggestion
### Validating CloudFormation files
```

If for any reason you need to edit the CloudFormation files,
be sure to lint them before using:

```

> $ pip3 install --upgrade taskcat cfn-lint

Reviewer: Not a blocker but it would be nice to provide a requirments.txt instead of an inline install like this.

$ taskcat lint && cfn-lint *cfn.yml
```

## I. Prologue: Ensure Source=Quilt Roles for Athena users

The initial Quilt deployment ships with two `Source=Custom` Roles that
automatically grant access to all buckets that have been added to Quilt.

> Unfortunately, at this time `Source=Custom` Roles are **not** compatible with accessing

Reviewer: ```suggestion
At this time `Source=Custom` Roles are incompatible with custom policies (including Athena policies).
```

> Athena.  Therefore, if you have not already, from the Quilt Admin Settings you

Reviewer: ```suggestion
Therefore before proceeding you must do the following in the Quilt Admin panel:
```

> must first:

Reviewer: ```suggestion
```

> 1. Create an empty `Source=Quilt` Role, e.g. "AthenaQuiltAccess"

Reviewer: ```suggestion
1. Create a new Quilt Role (`Source=Quilt`) , such as "AthenaQuiltAccess"
```

> 1. Create a new `Source=Quilt` Policy with access (usually "Read-write") to relevant S3 Buckets, and attach it to that Role

Reviewer: ```suggestion
1. Create a new Quilt Policy (`Source=Quilt`) with access, usually read and write to the query results bucket, any
buckets that your Athena tables will read from, and any buckets in the Quilt stack that you wish your users to
access.
```

1. Assign that Role to the Quilt Users that need to access Athena

> See the [Users and Roles](https://docs.quiltdata.com/catalog/admin)

Reviewer: ```suggestion
See [Users and Roles](https://docs.quiltdata.com/catalog/admin)
```

> documentation for more details about using Admin Settings to create and assign Roles and Policies.

Reviewer: ```suggestion
for more details on managing Roles and Policies.
```

> ## II. Upload Athena Configuration files to S3

Reviewer: ```suggestion
## II. Upload CloudFormation templates to S3
```

> There are two Athena templates in this directory,

Reviewer: ```suggestion
There are two templates in this directory.
```

> which must be uploaded directly to S3
> (as they are too large to read directly into CloudFormation)

Reviewer: ```suggestion
Both should be uploaded directly to S3 since they are larger than the AWS console allows for direct upload.
```

> *  `athena-cfn.yml` is only run once, and creates a results bucket and

Reviewer: ```suggestion
*  run `athena-cfn.yml` once to create a new workgroup and results bucket
```

> workgroup usable by the entire account (with appropriate permissions).

Reviewer: ```suggestion
```

> * `athena-bucket-cfn.yml` must be run *once for each Quilt bucket*

Reviewer: ```suggestion
* run`athena-bucket-cfn.yml` *once for each Quilt bucket* that you wish to SQL query
```

> that Athena should query. This creates bucket-specific tables and views which
> wrap the underlying Quilt metadata.

Reviewer: ```suggestion
The second template creates bucket-specific tables and views for querying object
and package metadata for any packages in the bucket in question.
```

> To install these in S3:

Reviewer: ```suggestion
### Upload templates to S3
```

> 1. Go to Amazon S3 -> Buckets  in the AWS Console for the region where most

Reviewer: ```suggestion
1. Go to AWS Console > S3 in the appropriate region.
```

> of your buckets are, e.g.:
> https://s3.console.aws.amazon.com/s3/buckets?region=us-east-1

Reviewer: ```suggestion
```

1. Choose any bucket you can read and write to
1. Create or navigate to an appropriate folder

> 1. Select Upload -> Add files

Reviewer: ```suggestion
1. Select Upload > Add files
```

1. Select both `athena-cfn.yml` and `athena-bucket-cfn.yml` from this repository
1. Click Upload
1. Command-click on each filename to open it in a new browser window
1. Remember or copy and paste the `https` 'Object URL' for each file

## III. Create Athena Workgroup

### A. Install `athena-cfn.yml` CloudFormation template

Note: If you store large amounts of data in multiple regions, you may need to
create and manage one such stack per _region_ (rather than per _account_).

> 1. Go to CloudFormation -> Stacks in the AWS Console for the region where most

Reviewer: ```suggestion
1. Go to AWS Console > CloudFormation > Stacks.
```

> of your buckets are, e.g.:
> https://us-east-1.console.aws.amazon.com/cloudformation/home?region=us-east-1#/stacks

Reviewer: ```suggestion
```

> 1. Click dropdown "Create stack" -> "With new resources (standard)"

Reviewer: For all such examples now quotes, leading capital and just `>` not ~~`=>`~~.

1. Select "Template is ready" (default)
1. Select "Amazon S3 URL" (default)
1. Paste in the `athena-cfn.yml` Object URL from above
1. Click Next
1. Enter "Stack name", e.g.: `athena-quilt`
1. Click Next
1. Click Next
1. Review then check "I acknowledge that AWS CloudFormation might create IAM resources"
1. Click "Create change set"
1. Wait until "Status = CREATE_COMPLETE" (may need to refresh using upper-right "cycle" icon)
1. Review Changes
1. Click "Execute" in upper right
1. Select "Roll back all stack resources" (default)
1. Click "Execute change set"

You need to wait until the `athena-cfn.yml` stack creation completes,
so you can use its Outputs for the following steps.

> ### B. Extend Quilt's Permission Boundary

Reviewer: Please link to main docs on this. Also need a separate PR that modifies the Athena page on docs.quiltdata.com and points here (if you haven't already).

Author: https://github.com/quiltdata/quilt/pull/3110

> 1. Go to the "Outputs" tab of the installed template

Reviewer: No quotes. And be more precise
```suggestion
1. Click the Outputs tab of the stack
```
